### PR TITLE
Feature/549/prune blocks

### PR DESCRIPTION
**Pull request by davikstone2**

### CODE REVIEW

File: saito-core/src/core/data/transaction.rs [hunk 1]
```diff
@@ -856,6 +856,21 @@ impl Transaction {
             return true;
         }
 
+        if self.transaction_type == TransactionType::SPV {
+            // Check if hash_for_signature contains the placeholder hash
+            if let Some(hash) = &self.hash_for_signature {
+                if *hash != [0; 32] {
```
Comment (SankaD) [inline saito-core/src/core/data/transaction.rs:862]: does this mean hash should be [0;32] ?

File: saito-core/src/core/data/merkle.rs [hunk 2]
```diff
@@ -64,66 +71,81 @@ impl MerkleTree {
...
+            while leaves.len() > 1 {
+                let left_leaf = leaves.pop_front().unwrap();
+                let right_leaf = leaves.pop_front().unwrap();
+
+                // Determine if the new node is an SPV node and if we should skip hashing.
+                let is_spv = left_leaf.is_spv && right_leaf.is_spv;
+                if is_spv {
+                    // If both are SPV transactions, push the left leaf only as the parent node.
+                    nodes.push_back(left_leaf);
+                    // Decide what to do with right_leaf - here we're discarding it
```
Comment (SankaD) [inline saito-core/src/core/data/merkle.rs:109]: is this TODO ?

File: saito-core/src/core/data/block.rs [hunk 3]
```diff
@@ -1916,6 +1944,42 @@ impl Block {
 
         transactions_valid
     }
+
+    // pub fn validate_spv_transaction(&self, spv_tx: &Transaction) -> bool {
```
Comment (SankaD) [inline saito-core/src/core/data/block.rs:1948]: remove if these are not needed

**Review by SankaD — CHANGES_REQUESTED** [review]

File: saito-core/src/core/data/block.rs [hunk 4]
```diff
@@ -1511,8 +1511,38 @@ impl Block {
             self.transactions.len()
         );
 
-        // TODO : prune transactions here
+        let mut i = 0;
+        let mut no_simplification_needed = false;
+        while !no_simplification_needed {
```
[the remark this answers is not in the document]
Comment (davikstone2) [reply saito-core/src/core/data/block.rs:1513]: there is a unit test "generate_lite_block_test" already for this

File: saito-core/Cargo.toml [hunk 5]
```diff
@@ -6,6 +6,7 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+sha2 = "0.9"
```
[the remark this answers is not in the document]
Comment (davikstone2) [reply saito-core/Cargo.toml]: removed

File: saito-core/src/core/data/merkle.rs (hunk 2 above)
Comment (davikstone2) [reply saito-core/src/core/data/merkle.rs:109]: updated

File: saito-core/src/core/data/transaction.rs (hunk 1 above)
Comment (davikstone2) [reply saito-core/src/core/data/transaction.rs]: updated

File: saito-rust/data/issuance/test/KEY [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (SankaD) [inline saito-rust/data/issuance/test/KEY:1]: what's this?

File: saito-rust/data/.gitignore [hunk 7]
```diff
@@ -1,5 +1,5 @@
 *.block
 *.sai
 /test/*
-/issuance/issuance
-/issuance/test/
+# /issuance/issuance
+# /issuance/test/
```
Comment (SankaD) [inline saito-rust/data/.gitignore:5]: why remove these ?

**Review by SankaD — CHANGES_REQUESTED** [review]

File: saito-core/src/core/data/block.rs [hunk 8]
```diff
@@ -642,6 +641,7 @@ impl Block {
     //
     // downgrade block
     //
+    #[ignore]
```
Comment (SankaD) [inline saito-core/src/core/data/block.rs:644]: what's this ?

File: saito-core/src/core/data/block.rs [hunk 9]
```diff
@@ -1486,15 +1488,9 @@ impl Block {
...
                     transaction_type: TransactionType::SPV,
                     txs_replacements: 1,
-                    signature: [
-                        tx.hash_for_signature.as_ref().unwrap().as_slice(),
-                        [0; 32].as_slice(),
-                    ]
-                    .concat()
-                    .try_into()
-                    .unwrap(),
+                    signature: tx.signature.clone(),
```
Comment (SankaD) [inline saito-core/src/core/data/block.rs:1491]: clone() is not required

File: saito-core/src/core/data/block.rs [hunk 10]
```diff
@@ -1486,15 +1488,9 @@ impl Block {
...
                     path: vec![],
-                    hash_for_signature: tx.hash_for_signature,
+                    hash_for_signature: tx.hash_for_signature.clone(),
```
Comment (SankaD) [inline saito-core/src/core/data/block.rs:1493]: clone is not required

File: saito-rust/data/issuance/test/issuance [hunk 11]
```diff
@@ -1,17 +1,8 @@
...
-99500	uqeqztfzneBcxFwthsyajKTECjciLnhg7gKt45S5DB95	Normal
-500	s8oFPjBX97NC2vbm9E5Kd2oHWUShuSTUuZwSB1U4wsPR	Normal
-99500	wdPpX51sNPZ6MhEm1BLFZNtmJozK7sFpfxEtHZAwdSzc	Normal
-99500	y8xgSTwgWH4zZszQyYdG4Mjkaa6nDEZfh2VzoC2ji7Tx	Normal
-99500	mzwgxik6L9zJkmZ2AA8dAjjyYmYVrDhNniysmtv6HYq9	Normal
-99500	oYrMgjc18RsaJfwahk8dyvyHaVoXPn6NtUnPvnVqFa7m	Normal
-99500	dSAuFiRw2EqfKxkUqJt3xKnVmvq8EaRHfpdK4dzAX7Wi	Normal
-99500	zeZmeHe5AM6Mt8aLkWF6D8rVTsHWZtJBK5WgtT5sh2mY	Normal
-99500	rJ9xm87Qzs5A4tpKf4hvqfNQZjQLmCQDsYJBcotsyzGe	Normal
+99500	i5Gh5B9K5TuC8ZN25Up3YGPvy5ikzZP1aNwGJ6WRGDLx	Normal
```
Comment (SankaD) [inline saito-rust/data/issuance/test/issuance:1]: add this file to gitignore please

**Review by SankaD — CHANGES_REQUESTED** [review]

**Review by SankaD — APPROVED** [review]

**Pull request merged**